# Kaggle Titanic Problem

In [1]:
import pandas as pd
import numpy as np

## Import the CSV files we need

In [2]:
df_train = pd.read_csv("data/train.csv")
df_test  = pd.read_csv("data/test.csv")

###### Take a look at the first 5 records of each dataset to check if the data's been imported correctly. Are there any problems here?

In [3]:
print("Training data header: ")
df_train.head(n=5)

Training data header: 


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [4]:
print("Test data header:")
df_test.head(n=5)

Test data header:


PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

###### Start with cleaning the training data. Can we pull up some basic stats for how complete the data is?

In [5]:
df_train.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [6]:
df_train.isnull().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

###### There are 177 missing ages - let's take a look at a few of these to see what's happening!

In [7]:
df_train[df_train['Age'].isnull()].head(n=5)

PassengerId  Survived  Pclass                           Name     Sex  Age  \
5             6         0       3               Moran, Mr. James    male  NaN   
17           18         1       2   Williams, Mr. Charles Eugene    male  NaN   
19           20         1       3        Masselmani, Mrs. Fatima  female  NaN   
26           27         0       3        Emir, Mr. Farred Chehab    male  NaN   
28           29         1       3  O'Dwyer, Miss. Ellen "Nellie"  female  NaN   

    SibSp  Parch  Ticket     Fare Cabin Embarked  
5       0      0  330877   8.4583   NaN        Q  
17      0      0  244373  13.0000   NaN        S  
19      0      0    2649   7.2250   NaN        C  
26      0      0    2631   7.2250   NaN        C  
28      0      0  330959   7.8792   NaN        Q

###### Cast objects to strings to make the data ready to use and visualise

In [8]:
df_train.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [9]:
df_train['Name'].astype(str)
df_train['Sex'].astype(str)
df_train['Cabin'].astype(str)
df_train['Embarked'].astype(str)

df_train.head(n=5)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

###### We now have data that's ready to visualise and use. Let's get onto feature engineering, shall we?

## Feature engineering

###### Create a feature engineering dataset. Drop Cabin, since it's mostly empty anyway and doesn't look too useful.

In [16]:
df_base = df_train[['PassengerId','Survived','Pclass','Sex','Age','SibSp','Parch','Ticket','Fare','Embarked']]

###### We could have just used df_base = df_train.drop(columns=['Cabin','Name']), but I like explicitely starting out with what columns we have.

###### Fill missing values in age with 0.

In [17]:
df_base['Age'].fillna(0, inplace=True)

df_base.describe()

df_base[df_base['Age'].isnull()]

Empty DataFrame
Columns: [PassengerId, Survived, Pclass, Sex, Age, SibSp, Parch, Ticket, Fare, Embarked]
Index: []

In [18]:
df_base['Embarked'].fillna('X', inplace=True)

df_base.describe()

df_base[df_base['Embarked'].isnull()]

Empty DataFrame
Columns: [PassengerId, Survived, Pclass, Sex, Age, SibSp, Parch, Ticket, Fare, Embarked]
Index: []

###### Categorise the Sex and Embarked variables to a number

In [19]:
df_base['Sex'].replace(['male','female'],[0,1],inplace=True)

df_base.head(n=5)

PassengerId  Survived  Pclass  Sex   Age  SibSp  Parch            Ticket  \
0            1         0       3    0  22.0      1      0         A/5 21171   
1            2         1       1    1  38.0      1      0          PC 17599   
2            3         1       3    1  26.0      0      0  STON/O2. 3101282   
3            4         1       1    1  35.0      1      0            113803   
4            5         0       3    0  35.0      0      0            373450   

      Fare Embarked  
0   7.2500        S  
1  71.2833        C  
2   7.9250        S  
3  53.1000        S  
4   8.0500        S

In [20]:
df_base['Embarked'].replace(['C','Q','S','X'],[1,2,3,-999],inplace=True)

df_base.head(n=5)

PassengerId  Survived  Pclass  Sex   Age  SibSp  Parch            Ticket  \
0            1         0       3    0  22.0      1      0         A/5 21171   
1            2         1       1    1  38.0      1      0          PC 17599   
2            3         1       3    1  26.0      0      0  STON/O2. 3101282   
3            4         1       1    1  35.0      1      0            113803   
4            5         0       3    0  35.0      0      0            373450   

      Fare  Embarked  
0   7.2500         3  
1  71.2833         1  
2   7.9250         3  
3  53.1000         3  
4   8.0500         3

In [21]:
df_base.loc[df_base['Embarked'] == -999]

PassengerId  Survived  Pclass  Sex   Age  SibSp  Parch  Ticket  Fare  \
61            62         1       1    1  38.0      0      0  113572  80.0   
829          830         1       1    1  62.0      0      0  113572  80.0   

     Embarked  
61       -999  
829      -999

###### Create a family column which is the sum of SibSp and Parch, to show how many family members someone has

In [ ]:
df_base['family'] = df_base['SibSp']